Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp

```diff
@@ -39,12 +39,110 @@ GStreamerVideoCapturer::GStreamerVideoCapturer(const char* sourceFactory)
 GstElement* GStreamerVideoCapturer::createConverter()
 {
     auto converter = gst_parse_bin_from_description("videoscale ! videoconvert ! videorate", TRUE, nullptr);
+#if USE(WESTEROS_SINK)
+// If camera supports only x-raw, westerossink expects encoded form. So pass through an h264 encoder and parser after probing platform.
+// If camera supports encoded formats, use a dummy identity element to skip converting to x-raw.
+    const char* codec = getPreferredCodec().c_str();
+    GST_INFO_OBJECT(m_pipeline.get(), "Preferred codec is %s ", codec);
+
+    if (strcmp (codec, "video/x-raw") == 0)
+    {
+        GstElementFactory *h264encoder = getEncoder ("video/x-h264");
+        GstElementFactory *h264parser = getParser ("video/x-h264");
+
+        char *videoInput = g_strdup_printf ("videoscale ! videoconvert ! videorate ! %s ! %s", GST_ELEMENT_NAME(h264encoder), GST_ELEMENT_NAME(h264parser));
+        converter = gst_parse_bin_from_description(videoInput, TRUE, nullptr);
+        m_caps = adoptGRef(gst_caps_new_empty_simple("video/x-h264"));
+    }
+    else
+    {
+        converter = gst_parse_bin_from_description("identity", TRUE, nullptr);
+        m_caps = adoptGRef(gst_caps_new_empty_simple(codec));
+    }
+#endif
 
     ASSERT(converter);
 
     return converter;
 }
 
+std::string GStreamerVideoCapturer::getPreferredCodec()
+{
+    GstCaps* caps = gst_device_get_caps(m_device.get());
+    const std::string videoCaps = gst_caps_to_string(caps);
+
+    if( videoCaps.find("video/x-av1") != std::string::npos )
+        return "video/x-av1";
+    else if( videoCaps.find("video/x-vp9") != std::string::npos )
+        return "video/x-vp9";
+    else if( videoCaps.find("video/x-vp8") != std::string::npos )
+        return "video/x-vp8";
+    else if( videoCaps.find("video/x-h265") != std::string::npos )
+        return "video/x-h265";
+    else if( videoCaps.find("video/x-h264") != std::string::npos )
+        return "video/x-h264";
+    else if( videoCaps.find("video/x-mpeg") != std::string::npos )
+        return "video/x-mpeg";
+    else if( videoCaps.find("video/x-raw") != std::string::npos )
+        return "video/x-raw";
+
+    return " ";
+}
+
+GstElementFactory* GStreamerVideoCapturer::getEncoder (const char* format)
+{
+    GList *encoder_list, *encoders;
+    GstCaps *caps_str;
```

Comment:
Use `GRefPtr<GstCaps>` if you think it's a good idea and prefer to avoid the `gst_caps_unref()` call later.

Reply:
Changed